src/pybind/mgr/dashboard/services/nvmeof_client.py

```diff
@@ -36,6 +36,19 @@ def __init__(self, gw_group: Optional[str] = None):
                     f'Unable to retrieve the gateway info: {e}'
                 )
 
+            # While creating listener need to direct request to the gateway
+            # address where listener is supposed to be added.
+            if traddr:
+                gateways_info = NvmeofGatewaysConfig.get_gateways_config()
+                matched_instance = next(
+                    filter(lambda instance: traddr in instance['service_url'],
+                        (instance for instances in gateways_info['gateways'].values() for instance in instances)),
+                    None
+                )
```

Comment:
Whether you use `map/filter` or comprehensions, let's try to make thing easy to understand:
```suggestion
                matched_instance = next(
                    (
                        gateway
                        for gateways in gateways_info["gateways"].values()
                        for gateway in gateways
                        if traddr in instance["service_url"]
                    ),
                    None,
                )
```
That said, this type of patterns are repeated many times in our code, and IMHO that's because instead of mapping input data to our data classes before operating with that, we directly process external data structures and end up having "hybrid-procedural" code instead of true OOP code.

If we were using Dashboard native abstractions, this would look like (`gateways` could be our list of our custom `NvmeofGateway` classes):
```suggestion
                matched_instance = filter(gateways, lambda gw: gw.taddr == taddr)
```
Or, if we model our data structures based on the queries we do (now `Gateways` is a class that acts as a querier/factory between NVMe-oF data and Dashboard data):
```suggestion
                matched_instance = Gateways.get_by_taddr(taddr, None)
```

Reply:
The main goal of this is encapsulation and reuse.

Reply:
I agree with most of that, but being a bug fix I did not touch many functions/implementations although they can be improved.

We can do that gradually. For this PR we dont need to do that, so lets focus on the bug fix and if that being fixed rightly. I am noting down your suggestions, thanks

Reply:
I will fix the readability part 👍🏿

Reply:
Tracker: https://tracker.ceph.com/issues/68273